flink/src/main/java/org/apache/iceberg/flink/sink/FlinkManifest.java

```diff
@@ -0,0 +1,105 @@
+import org.apache.iceberg.io.LocationProvider;
+import org.apache.iceberg.io.OutputFile;
+import org.apache.iceberg.io.OutputFileFactory;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+class FlinkManifest {
+  private static final int ICEBERG_FORMAT_VERSION = 2;
+  private static final Long DUMMY_SNAPSHOT_ID = 0L;
+
+  private final OutputFile outputFile;
+  private final PartitionSpec spec;
+
+  private FlinkManifest(OutputFile outputFile, PartitionSpec spec) {
+    this.outputFile = outputFile;
+    this.spec = spec;
+  }
+
+  ManifestFile write(List<DataFile> dataFiles) throws IOException {
+    ManifestWriter<DataFile> writer = ManifestFiles.write(ICEBERG_FORMAT_VERSION, spec, outputFile, DUMMY_SNAPSHOT_ID);
+    try (ManifestWriter<DataFile> closeableWriter = writer) {
+      closeableWriter.addAll(dataFiles);
+    }
+    return writer.toManifestFile();
+  }
+
+  static List<DataFile> read(ManifestFile manifestFile, FileIO io) throws IOException {
+    try (CloseableIterable<DataFile> dataFiles = ManifestFiles.read(manifestFile, io).project(ManifestFile.schema())) {
+      return Lists.newArrayList(dataFiles);
+    }
+  }
+
+  static FlinkManifestFactory createFactory(Table table, String flinkJobId, int partitionId, long taskId) {
+    return new FlinkManifestFactory(table.location(), table.spec(), FileFormat.AVRO, table.locationProvider(),
+        table.io(), table.encryption(), flinkJobId, partitionId, taskId);
+  }
+
+  static class FlinkManifestFactory {
+    private final String tableLocation;
+    private final PartitionSpec spec;
+    private final FileFormat format;
+    private final OutputFileFactory outputFileFactory;
+    private final int partitionId;
+    private final long taskId;
+    private final String flinkJobId;
+    private final AtomicInteger fileCount = new AtomicInteger(0);
+
+    FlinkManifestFactory(String tableLocation, PartitionSpec spec, FileFormat format, LocationProvider locations,
+                         FileIO io, EncryptionManager encryptionManager, String flinkJobId, int partitionId,
+                         long taskId) {
+      this.tableLocation = tableLocation;
+      this.spec = spec;
+      this.format = format;
+      this.flinkJobId = flinkJobId;
+      this.partitionId = partitionId;
+      this.taskId = taskId;
+      this.outputFileFactory = new OutputFileFactory(spec, format, locations, io,
+          encryptionManager, partitionId, taskId);
+    }
+
+    private String generateRelativeFilePath(long checkpointId) {
+      return format.addExtension(
+          String.format("%s/flink-manifest/%s-%05d-%d-%d-%05d", tableLocation, flinkJobId, partitionId, taskId,
+              checkpointId, fileCount.incrementAndGet()));
+    }
+
+    FlinkManifest create(long checkpointId) {
+      String relativeFilePath = generateRelativeFilePath(checkpointId);
+      return new FlinkManifest(outputFileFactory.newOutputFile(relativeFilePath).encryptingOutputFile(), spec);
```

Comment:
This should use `TableOperations.metadataFileLocation(filename)` to create the full location, and should use `FileIO` directly to create the output file.

The output file factory is used for data files and it creates paths using the location provider. The location provider is only used for data files. For metadata files, `TableOperations.metadataFileLocation` is used.

Also, there is no key metadata tracking for metadata files, so using the encryption manager would produce a file that is either plaintext or unreadable because the key metadata is lost.